File: 2023-08-25/b/main.cpp
#include <iostream>
#include <vector>
#include <map>

class Frog 
{
private:
  int hop=0;
  int pos=0;
  int wei=0;

public:
  Frog () {};
  Frog (int hop) : hop(hop) {}

  void describe () const
  {
    std::cout << "hop: " << this->hop << ", pos: " << this->pos << ", wei: " << this->wei << std::endl; 
  }
    
  void increaseWeight()
  {
    this->wei++; 
  }

  int jump()
  {
    this->pos += this->hop;
    return this->pos;
  }

  int getWeight()
  {
    return this->wei;    
  }

  int getPosition()
  {
    return this->pos;
  }

};

int main()
{
  int N; std::cin >> N;

  while(N)
  {
    N--; 
    int n; std::cin >> n;


    std::map< int, Frog > frogMap;


    for(int i=0; i<n; i++) 
    {
      int hop; std::cin >> hop;
      if( !frogMap.count(hop) )
        frogMap[hop] = Frog(hop);
      frogMap[hop].increaseWeight();
    }


    int *arr = new int[n+1];
    for(int i=0; i<n+1; i++)
      arr[i] = 0;

    int maxWeight = 0;

    for( auto& pair : frogMap )
    {
      Frog& frog = pair.second;      
      while(true)
      {
        int pos = frog.jump();
        if(pos > n)
          break;
        arr[pos] += frog.getWeight();
        if(maxWeight < arr[pos])
          maxWeight = arr[pos];        
      }
    }


    delete[] arr;


    std::cout << maxWeight << std::endl;
  }
}

File: 2023-08-25/b/main.py
class Frog:
    def __init__(self, hop):
        self.hop = hop
        self.pos = 0
        self.wei = 1


    def position(self):
        return self.pos


    def increaseWeight(self):
        self.wei += 1


    def weight(self):
        return self.wei


    def jump(self):
        self.pos += self.hop


    def __repr__(self):
        return 'Frog H {} P {} W {}'.format(self.hop, self.pos, self.wei)


if __name__ == '__main__':
    N = int( input() )
    for i in range(N):
        n = int( input() )
        frogs = {}
        
        jumpPositions = [0]*(n+1)
        maxJump = 0

        for hop in list(map(int, input().split())):
            if hop not in frogs:
                frogs[hop] = Frog(hop=hop)
                continue
            frogs[hop].increaseWeight()

        frogsList = list( frogs.values() )

        for frog in frogsList:
            while True:
                frog.jump()
                if frog.pos > n:
                    break
                jumpPositions[frog.pos] += frog.wei

                if maxJump < jumpPositions[frog.pos]:
                    maxJump = jumpPositions[frog.pos]

        print(maxJump)



